### PR TITLE
Added KMP substring search

### PR DESCRIPTION
**Pull request by GervinFung**
Good day sir, I took some time and have implemented KMP sub-string search as I think it's beneficial to learn why advanced searching algorithm is preferred and how it works and why it has lower time complexity than naive/brute-force algorithm

I look forward to your reply and maybe the merge of my code to the main branch!
Thank you
Regards，
Gervin Fung

### CODE REVIEW

**Comment by threelittle87** [comment]
Nice work

File: src/strings/kmp_substring_search.jl [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        end
+    end
+
+    return j == substring_length + 1 ? i - j + 1 : NO_SUBSTRING_INDEX
+end
+
+# optional function that returns boolean if string does contain the sub-string given
+function ContainSubstringWithKMP(string::String, sub_string::String, ignore_case::Bool)::Bool
+    return GetIndexWithKMP(string, sub_string, ignore_case) != NO_SUBSTRING_INDEX
+end
```
Comment (Panquesito7) [inline src/strings/kmp_substring_search.jl:100]: Missing endline.
```suggestion
end

```

File: src/strings/kmp_substring_search.jl [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+Question
+1. Why KMP instead of naive search?
+- Given n = length of string , m = length of string P
+- Then naive search will have time complexity O(mn) to find all occurrences of pattern P in S
+- However, knuth-morris-pratt (KMP) algorithm will have tim complexity of O(m+n) to find all occurrences of pattern P in S
+- Hence, KMP is more efficient than naive approach
+
+
+Example
```
Comment (Panquesito7) [inline src/strings/kmp_substring_search.jl:16]:
```suggestion
Example

```

File: src/strings/kmp_substring_search.jl [hunk 3]
```diff
@@ -0,0 +1,100 @@
+"""
+    GetIndexWithKMP(string::String, sub_string::String, ignore_case::Bool)::Int
+
+Find if a string contain a given sub string with KMP Search, as an explanation might be too lengthy
+As such, a detailed explanation can be found at https://towardsdatascience.com/pattern-search-with-the-knuth-morris-pratt-kmp-algorithm-8562407dba5b
+
+
+Question
```
Comment (Panquesito7) [inline src/strings/kmp_substring_search.jl:8]: I think it should be like this here, as you are providing an answer.
```suggestion
Questions and answers

```

File: src/strings/kmp_substring_search.jl [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+1. Why KMP instead of naive search?
+- Given n = length of string , m = length of string P
+- Then naive search will have time complexity O(mn) to find all occurrences of pattern P in S
```
Comment (Panquesito7) [inline src/strings/kmp_substring_search.jl:11]:
```suggestion
- Then, the naive search will have time complexity O(mn) to find all occurrences of pattern P in S
```

File: src/strings/kmp_substring_search.jl [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+- However, knuth-morris-pratt (KMP) algorithm will have tim complexity of O(m+n) to find all occurrences of pattern P in S
+- Hence, KMP is more efficient than naive approach
```
Comment (Panquesito7) [inline src/strings/kmp_substring_search.jl:13]:
```suggestion
- Hence, KMP is more efficient than the naive approach
```

Comment (GervinFung) [reply src/strings/kmp_substring_search.jl]: Thank you for your reply @Panquesito7, I had already made the changes as you requested. Do inform me if there's any changes needed

File: src/strings/kmp_substring_search.jl [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+```julia
+GetIndexWithKMP("Hello I am Gervin", "Hello I am Gervin", false) # returns 1
+GetIndexWithKMP("ABABDABACDABABCABAB", "ABABCABAB", false) # returns 11
+GetIndexWithKMP("SeEms Like IgNOrE CaSe Work", "seems like ignore case work", true) # returns 1
+```
+
+Contributed By:- [Gervin Fung](https://github.com/GervinFung)
+
+
```
Comment (Panquesito7) [inline src/strings/kmp_substring_search.jl:26]:
```suggestion
```

File: src/strings/kmp_substring_search.jl [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+Note: This function will also allow ignore case also
```
Comment (Panquesito7) [inline src/strings/kmp_substring_search.jl:27]:
```suggestion
Note: This function will also allow ignoring cases also
```

Comment (GervinFung) [reply src/strings/kmp_substring_search.jl]: Thank you for pointing out the grammar mistake in the documentation! I had already changed it as well.

File: src/strings/kmp_substring_search.jl [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+Example
+
+```julia
+GetIndexWithKMP("Hello I am Gervin", "Hello I am Gervin", false) # returns 1
+GetIndexWithKMP("ABABDABACDABABCABAB", "ABABCABAB", false) # returns 11
+GetIndexWithKMP("SeEms Like IgNOrE CaSe Work", "seems like ignore case work", true) # returns 1
+```
+
+Contributed By:- [Gervin Fung](https://github.com/GervinFung)
+
```
Comment (Panquesito7) [inline src/strings/kmp_substring_search.jl:25]:
```suggestion
```

Comment (GervinFung) [reply src/strings/kmp_substring_search.jl]: Hi @Panquesito7 , thanks for your quick reply. I already made the changes as well

**Review by Azzaare — APPROVED** [review]
The PR looks good

Although it is not mandatory for the code to work, it is preferable to follow the code format adopted by other people. In particular, beside constructors, functions usually follow the snake_case convention in Julia. I will correct it later (as it is not the only PR that introduced CamelCase for function).

**Pull request merged**